### PR TITLE
[Enhancement] reuse json parser to speed up json parsing

### LINKED ISSUE #4556: reuse simdjson parser and buffer to speed up loading json
## Enhancement
The `simdjson::ondemand::parser` and its buffer is recommended to be reused to speed up the parsing.
Especially in routine load, we parse the json message from kafka one by one, in which constructing `simdjson::ondemand::parser` and allocating buffer brings significant overhead.

### PR DESCRIPTION
**Pull request by rickif**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #4556 

## Problem Summary(Required) ：

This PR reuse the json parser and buffer to get better performance, which is recommended in 
https://github.com/simdjson/simdjson/blob/master/doc/performance.md

### CODE REVIEW

**Review by murphyatwork — DISMISSED** [review]

**Comment by rickif** [comment]
## Benchmark
Type: Routine Load
Source: message in JSON format; 300 bytes per message; 1 partition in Kafka
StarRocks Config: 3 BE with 3 replica

| version | output|
| main|32.51MB/s |
| perf/reuse-parser| 95.88MB/s|

**Comment by rickif** [comment]
run starrocks_be_unittest

File: be/src/runtime/stream_load/stream_load_pipe.h [hunk 1]
```diff
@@ -86,28 +86,32 @@ class StreamLoadPipe : public MessageBodySink, public FileReader {
     // If _total_length == -1, this should be a Kafka routine load task,
     // just get the next buffer directly from the buffer queue, because one buffer contains a complete piece of data.
     // Otherwise, this should be a stream load task that needs to read the specified amount of data.
-    Status read_one_message(std::unique_ptr<uint8_t[]>* data, size_t* length, size_t padding) override {
+    Status read_one_message(std::unique_ptr<uint8_t[]>* buf, size_t* buf_cap, size_t* buf_sz, size_t padding) override {
```
Comment (chaoyli) [inline be/src/runtime/stream_load/stream_load_pipe.h:89]: What's the purpose of buf_cap?

File: be/src/exec/vectorized/json_scanner.h [hunk 2]
```diff
@@ -101,12 +101,15 @@ class JsonReader {
...
-    int _next_line;
-    int _total_lines;
     bool _closed;
     std::vector<SlotDescriptor*> _slot_descs;
 
-    std::unique_ptr<uint8_t[]> _json_binary_ptr;
+    // For performance reason, the simdjson parser should be reused over several files.
+    //https://github.com/simdjson/simdjson/blob/master/doc/performance.md
+    simdjson::ondemand::parser _simdjson_parser;
+    std::unique_ptr<uint8_t[]> _parser_buf;
```
Comment (chaoyli) [inline be/src/exec/vectorized/json_scanner.h:110]: Will every load have a _parse_buf?
Does it will consumes much memory for a high concurrency load?

**Review by chaoyli — CHANGES_REQUESTED** [review]

File: be/src/runtime/stream_load/stream_load_pipe.h (hunk 1 above)
Comment (rickif) [reply be/src/runtime/stream_load/stream_load_pipe.h:89]: `buf_cap` denotes the capacity of buf. When `buf_cap` is bigger than payload size, `buf` could be reused and the actual size of payload would be returned in `buf_sz`. When `buf_cap` is smaller than payload size, `buf` should grow to required size and the original one should be released.

To make this clear, more comment is added.

File: be/src/exec/vectorized/json_scanner.h (hunk 2 above)
Comment (rickif) [reply be/src/exec/vectorized/json_scanner.h:110]: The previous implement is one buffer for one message. The new implement is one buffer for one load, which would not bring extra memory consuming.

**Comment by rickif** [comment]
@mergifyio rebase

File: be/src/exec/vectorized/json_scanner.h (hunk 2 above)
Comment (chaoyli) [reply be/src/exec/vectorized/json_scanner.h:110]: How much memory will a load consume?

**Review by chaoyli — APPROVED** [review]

**Review by Astralidea — APPROVED** [review]

**Pull request merged**